### PR TITLE
add invokable support to strippedRequest

### PR DESCRIPTION
**Pull request by tabacitu**
### BEFORE - What was wrong? What was happening before this PR?

See #4503 

You couldn't do `php artisan config:cache` if you defined a custom `strippedRequest`.

### AFTER - What is happening after this PR?

You can define a custom `strippedRequest` AND have it cached, by making it an **invokable class** instead of a **closure**. So instead of defining your logic right there in the config file, you create a separate file to hold that logic:

```php
<?php

namespace App\Http\Requests;

use Illuminate\Http\Request;

class StripBackpackRequest
{
    public function __invoke(Request $request)
    {
        return $request->except('_token', '_method', '_http_referrer', '_current_tab', '_save_action');
    }
}

```

### How did you achieve that, in technical terms?

Check if invokable, if so... invoke 😀

### Is it a breaking change?

Non-breaking.

### How can we test the before & after?

Before: 
- uncomment the closure in `config/backpack/operations/update.php`;
- if you try `php artisan config:cache` it will error;

After: 
- create the file above, then use that instead of the closure in your `config/backpack/operations/update.php`;
- if you try `php artisan config:cache` it should work just fine;

## TODO

- [ ] add documentation for the above... and maybe even for the [alternative solutions outlined here](https://github.com/Laravel-Backpack/CRUD/issues/4503#issuecomment-1177393044);

### CODE REVIEW

**Comment by pxpm** [comment]
I think this is too much for the problem we have to solve. Unless you specifically want that this config still viable, otherwise we should just remove the config at all, no need for it anymore. 

`strippedRequest` becomes backpack by default (currently false on config), or Dev uses the ServiceProvider to Provide (who have guessed), different configuration for this setting. (an array or a closure is what makes sense). 

~~~~~ that was old me thinking

New Me: Abstracting this to ServiceProvider might have the downside of not knowing where this config (strippedRequest) is comming from. Also would force the dev to go to ServiceProvider when they want to define it as array ...

So yeah, I think this is probably the best solution in the end. 

Will give it a try.

File: src/config/backpack/operations/update.php [hunk 1]
```diff
@@ -37,13 +37,15 @@
...
+ * - invokable class - eg. App\Http\Requests\StripBackpackRequest::class.
  *
  * @param  \Illuminate\Http\Request  $request
  * @return array
  */
+    // 'strippedRequest' => App\Http\Requests\StripBackpackRequest::class,
     // 'strippedRequest' => (function ($request) {
     //     return $request->except('_token', '_method', '_http_referrer', '_current_tab', '_save_action');
-    // }),
+    //  }),
```
[the remark this answers is not in the document]
Comment (tabacitu) [reply src/config/backpack/operations/update.php:50]: What?! WHAT?! But... why? 

hmmm_no_idea


1) Why does that work? 😅 please ELI5 😅
2) If it does work, I say we should _default_ to running that, instead of the closure. 
3) Does this still get cached? I would suspect not... but then again... we don't need this one cached, so that should be ok. The invokable class doesn't get cached either (just its name), so it's the same thing in that respect.
4) If this work... do we really need the invokable?! I don't think so.

[the remark this answers is not in the document]
Comment (maurohmartinez) [reply src/config/backpack/operations/update.php:50]: I have no idea why this even works... tbh
I would say the invokable seems a bit less hacky than this one 😆

[the remark this answers is not in the document]
Comment (pxpm) [reply src/config/backpack/operations/update.php:50]: I guess this works because it auto-executes the closure (runs the function) and `strippedRequest` get's an array value that can be serialized. Not sure it will work (it can by-pass the `config:cache`) but does it WORK ?

[the remark this answers is not in the document]
Comment (tabacitu) [reply src/config/backpack/operations/update.php:50]: Ah right. So basically it gets run (and calculated) one time, then the array is cached. So the `strippedRequest` would forever be the `request` from the `php artisan config:cache` call. _No bueno._ Good catch @pxpm . Nice try @maurohmartinez 😅

Yup, let's go with invokable then 👍

**Comment by tabacitu** [comment]
@pxpm wanna have a last pass at this?

File: src/app/Library/CrudPanel/Traits/Fields.php [hunk 2]
```diff
@@ -479,6 +479,15 @@ public function getStrippedSaveRequest($request)
     {
         $setting = $this->getOperationSetting('strippedRequest');
 
+        // if an invokable class was passed
+        // eg. \App\Http\Requests\BackpackStrippedRequest
+        if (class_exists($setting)) {
+            $setting = new $setting();
+
+            return $setting($request);
```
Comment (pxpm) [inline src/app/Library/CrudPanel/Traits/Fields.php:487]:
```suggestion
            return is_callable($setting) ? $setting($request) : abort(500, get_class($setting) . ' is not invokable.';
```

File: src/config/backpack/operations/update.php [hunk 3]
```diff
@@ -37,13 +37,15 @@
 
 /**
  * Before saving the entry, how would you like the request to be stripped?
- * - false - fall back to Backpack's default (ONLY save inputs that have fields)
- * - closure - process your own request (example removes all inputs that begin with underscode).
+ * - false - use Backpack's default (ONLY save inputs that have fields)
+ * - closure - custom request processing (quick, but will stop config caching)
```
Comment (pxpm) [inline src/config/backpack/operations/update.php:41]: same comment as above.

File: src/config/backpack/operations/create.php [hunk 4]
```diff
@@ -37,13 +37,15 @@
 
 /**
  * Before saving the entry, how would you like the request to be stripped?
- * - false - fall back to Backpack's default (ONLY save inputs that have fields)
- * - closure - process your own request (example removes all inputs that begin with underscode).
+ * - false - use Backpack's default (ONLY save inputs that have fields)
+ * - closure - custom request processing (quick, but will stop config caching)
```
[the remark this answers is not in the document]
Comment (tabacitu) [reply src/config/backpack/operations/create.php:41]: 😅 I agree. But we can't remove it now, it's been a feature for 4 months already. Gotta keep it until our next release, otherwise it's just confusing for someone who's used it before...

File: src/app/Library/CrudPanel/Traits/Fields.php (hunk 2 above)
Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Fields.php:487]: Aaah yes, good catch @pxpm ! Better to help the dev out with a clear message 👏👏👏 But you forgot to close the parantheses there:

```suggestion
            return is_callable($setting) ? $setting($request) : abort(500, get_class($setting) . ' is not invokable.');
```

File: src/config/backpack/operations/create.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (tabacitu) [reply src/config/backpack/operations/create.php]: Future-me here. 

Damn it! Yeah, I guess you're right @pxpm ... we shouldn't keep instruct people on how to do something WRONG... 😔 we'd be making more problems here...

Ok I removed that instruction. Good catch!

[the remark this answers is not in the document]
Comment (pxpm) [reply src/config/backpack/operations/create.php]: I don't think we **can't** remove it. 

- New users will not get it - fine
- Old users that might have a closure setup ... will still have the closure setup and working - fine
- Old users starting a new project ... read the docblock and adapt? - fine
- Old users starting a new project ... copy from old project? works as previously. - fine 

We are leaving something bad because probably 10% (I might be exagerating here) of devs that overwrote this in the last 4 months may be confused ? It still does not sell to me.

I guess 99% of the 10% that configures this setting `strippedRequest` do it on a per crud basis, because only in crud X and Z they have some extra field they need to handle. 

I don't think the closure on config was used that many times.

This said I rest my case, if this was not enough to change your mind, please mark the conversation as solved and let's move on with our lives 👍

File: src/config/backpack/operations/create.php [hunk 5]
```diff
@@ -35,15 +35,8 @@
...
- *
- * @param  \Illuminate\Http\Request  $request
- * @return array
- */
-    // 'strippedRequest' => (function ($request) {
-    //     return $request->except('_token', '_method', '_http_referrer', '_current_tab', '_save_action');
-    // }),
+    // Before saving the entry, how would you like the request to be stripped?
+    //  - false - use Backpack's default (ONLY save inputs that have fields)
+    //  - invokable class - custom stripping (eg. App\Http\Requests\StripBackpackRequest::class)
```
Comment (tabacitu) [inline src/config/backpack/operations/create.php:40]:
```suggestion
    //  - invokable class - custom stripping (the return should be an array with input names) 
```

File: src/config/backpack/operations/update.php [hunk 6]
```diff
@@ -35,15 +35,8 @@
...
- *
- * @param  \Illuminate\Http\Request  $request
- * @return array
- */
-    // 'strippedRequest' => (function ($request) {
-    //     return $request->except('_token', '_method', '_http_referrer', '_current_tab', '_save_action');
-    // }),
+    // Before saving the entry, how would you like the request to be stripped?
+    //  - false - use Backpack's default (ONLY save inputs that have fields)
+    //  - invokable class - custom stripping (eg. App\Http\Requests\StripBackpackRequest::class)
```
Comment (tabacitu) [inline src/config/backpack/operations/update.php:40]:
```suggestion
    //  - invokable class - custom stripping (the return should be an array with input names) 
```

File: src/config/backpack/operations/create.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (tabacitu) [reply src/config/backpack/operations/create.php]: No, you're right. I changed my mind and hit "send" 1 min before you did 😅 sorry for the trouble

[the remark this answers is not in the document]
Comment (pxpm) [reply src/config/backpack/operations/create.php]: Nice to know I changed your mind while I was thinking and writting about it. We should try that again, maybe this was not just a coincidence 🥳

**Comment by pxpm** [comment]
https://github.com/Laravel-Backpack/docs/pull/360 - working on them, will approve when finished

**Review by pxpm — APPROVED** [review]
way to go. please review docs and merge when ok!

Cheers
